File: two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // Create a hash map to store the numbers and their indices
    let numMap = new Map();
    
    // Iterate through the array
    for (let i = 0; i < nums.length; i++) {
        let complement = target - nums[i];
        
        // Check if the complement exists in the map
        if (numMap.has(complement)) {
            // Return the indices of the current number and the complement
            return [numMap.get(complement), i];
        }
        
        // Add the current number and its index to the map
        numMap.set(nums[i], i);
    }
    
    // If no solution is found (should not happen according to the problem constraints)
    return [];
};

// Example usage:
console.log(twoSum([2, 7, 11, 15], 9)); // Output: [0, 1]
console.log(twoSum([3, 2, 4], 6)); // Output: [1, 2]
console.log(twoSum([3, 3], 6)); // Output: [0, 1]

File: two-sum/two-sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  num_map = {} # Create a hash to store the numbers and their indices
  
  # Iterate through the array
  nums.each_with_index do |num, index|
    complement = target - num
    
    # Check if the complement exists in the hash
    if num_map.has_key?(complement)
      # Return the indices of the current number and the complement
      return [num_map[complement], index]
    end
    
    # Add the current number and its index to the hash
    num_map[num] = index
  end
  
  # If no solution is found (should not happen according to the problem constraints)
  return []
end

# Example usage:
puts two_sum([2, 7, 11, 15], 9).inspect # Output: [0, 1]
puts two_sum([3, 2, 4], 6).inspect     # Output: [1, 2]
puts two_sum([3, 3], 6).inspect        # Output: [0, 1]
